Add Mixxx Control documentation to manual (#151)
mixxxdj/manual: source/chapters/advanced_topics.rst

For more information, go to `<https://mixxx.org/wiki/doku.php/midi_scripting>`_
and `<https://mixxx.org/wiki/doku.php/hid_mapping_format>`_.

Mixxx Controls
--------------

See :ref:`controlindex` for a full list.

ControlPotMeter controls
~~~~~~~~~~~~~~~~~~~~~~~~

The following extensions add some features to ControlPotMeter controls
(volume, crossfader, ...). Use in conjunction with ``[Channel*N*]``,
``[Sampler*N*]``, ``[Master]`` ... groups. Please note, this doesn't work in
JavaScript files, so you can not do, for example:

.. code-block:: javascript

>    engine.setValue('MyController', 'keylock_toggle')

Reviewer: I made the note more clear on the wiki, but I don't have time to copy it over here and apply formatting for a proper suggestion.. :\
for detail please see https://github.com/mixxxdj/mixxx/wiki/Mixxxcontrols/_compare/fcef2c24516d49f7dd8746868db5931bb5b5cfd9...fdf8a2fad2558a43fe4341913802ec130d7b8c3c

   :range: binary
   :feedback: Outro start marker changes on waveform.

   .. deprecated:: v2.3.0
      This control has been removed.

.. mixxx:control:: [ChannelN],outro_start_enabled

   1 if outro start cue is set, (position is not -1), 0 otherwise.

   :range: binary, read-only
   :feedback: Outro start button lights up.

   .. deprecated:: v2.3.0
      This control has been removed.

.. mixxx:control:: [ChannelN],outro_start_position

   The position of the outro start in samples, -1 if not set.

   :range: positive integer
   :feedback: Outro start marker changes on waveform.

   .. deprecated:: v2.3.0
      This control has been removed.

.. mixxx:control:: [ChannelN],outro_start_set

   Set outro start to the current play position. If outro start was previously set, it is moved to the new position.

   :range: binary
   :feedback: Outro start marker changes on waveform.

   .. deprecated:: v2.3.0
      This control has been removed.

.. mixxx:control:: [ChannelN],passthrough

   Connects the vinyl control input for vinyl control on that deck to the channel output. Allows to mix external media into DJ sets.

   :range: binary

>    :feedback: GUI control currently missing FIXME

Reviewer: ```suggestion
   :feedback: PASSTHROUGH label in the track overview, Passthrough button
```

.. mixxx:control:: [ChannelN],waveform_zoom

   Zooms the waveform to look ahead or back as needed.

   :range: 1.0 - 10.0
   :feedback: Waveform zoom buttons

   .. versionadded:: v1.11.0

.. mixxx:control:: [ChannelN],waveform_zoom_up

   Waveform Zoom Out

   :range: ?
   :feedback: Waveform zoom buttons

   .. versionadded:: v1.11.0

.. mixxx:control:: [ChannelN],waveform_zoom_down

   Waveform Zoom In

   :range: ?
   :feedback: Waveform zoom buttons

   .. versionadded:: v1.11.0

.. mixxx:control:: [ChannelN],waveform_zoom_set_default

   Return to default waveform zoom level

   :range: ?
   :feedback: Waveform zoom buttons

   .. versionadded:: v1.11.0

.. mixxx:control:: [ChannelN],wheel

   Affects relative play speed & direction persistently (additive offset & must manually be undone)

   :range: -3.0..3.0
   :feedback: Waveform

The Sampler controls
~~~~~~~~~~~~~~~~~~~~

Sample decks ("samplers") in Mixxx are identical to decks, they simply
have a different purpose (playing samples). Any control listed above for
``[ChannelN]`` will work for a sampler, just replace ``[ChannelN]`` with
``[SamplerN]``.

.. mixxx:control:: [SamplerN],[Samplers],show_samplers

> .. mixxx:control:: [SamplerN],[Sampler],SaveSamplerBank

Reviewer: ```suggestion
.. mixxx:control:: [Sampler],SaveSamplerBank
```

   :range: binary
   :feedback: Shows Sampler bank(s)

   Save sampler configuration. Make currently loaded tracks in samplers instantly available at a later point.

   :range: binary
   :feedback: Opens file dialog. Configuration file can be named and saved.

.. mixxx:control:: [SamplerN],[Sampler],LoadSamplerBank

   Load saved sampler configuration file and add tracks to the available samplers.

   :range: binary
   :feedback: Opens file dialog. Select configuration file.

Auxiliary controls
~~~~~~~~~~~~~~~~~~

You can map audio interface's inputs to mixxx's auxiliary input channels and
connect external audio source to it (cellphone, mp3 player). Then you can use
your midi controller to control its volume and some other parameters

> (orientation, gain, volume), apply effects and use listen function.

Reviewer: ```suggestion
(orientation, gain, volume), apply effects and use prelisten function.
A lot of controls from the ``[ChannelN]`` list can be used, too, except for example EQ controls and obviously controls that relate to beats, playback direction etc.
```
```

~~~~~~~~~~~~~~~~~~~~~

Preview decks in Mixxx are identical to regular decks. Any control listed above
for ``[ChannelN]`` will work for a preview deck, just replace ``[ChannelN]``
with ``[PreviewDeckN]``.

The ``[VinylControl]`` group
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. mixxx:control:: [VinylControl],Toggle

   Moves control by a vinyl control signal from one deck to another if using the single deck vinyl control (VC) feature.

   :range: binary
   :feedback: If VC isn't enabled on any decks, enable it on the first one we're receiving samples for. If VC is enabled on a single (exclusive) deck, and another deck is setup to receive samples, disable it on the former deck and enable it on the next eligible deck (ordered by deck number). If VC is enabled on multiple decks, don't do anything.

   .. versionadded:: v1.10.0

The ``[MicrophoneN]`` group
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Below, N=2 up to the number of active microphones. e.g ``[Microphone2]``. For Microphone 1, just use ``[Microphone]``.

.. mixxx:control:: [MicrophoneN],enabled

   1 if a microphone input is enabled, 0 if not.

   :range: binary
   :feedback: Microphone is enabled.

   .. versionadded:: v1.10.0

> .. mixxx:control:: [MicrophoneN],orientation
> 
>    Set microphone orientation,
> 
>    :range:
>       ===== ===================================
>       Value Meaning
>       ===== ===================================
>       0     Left side of crossfader
>       1     Center (not affected by crossfader)
>       2     Right side of crossfader
>       ===== ===================================
>    :feedback: N/A
> 
>    .. versionadded:: v1.10.0
> 
> 

Reviewer: The entire section should be removed.
Controls are being created indeed but not processed in the mixer, that's why we don't have orientation controls in the GUI. I just filed https://bugs.launchpad.net/mixxx/+bug/1899639
```suggestion
```

Author: I added a deprecation note, so that people who find it in the developer tools don't get confused.

   Adjusts the microphone volume fader

   :range: default
   :feedback: Microphone volume fader changes

   .. versionadded:: v1.10.0

.. mixxx:control:: [MicrophoneN],pregain

   Adjusts the gain of the mic input

   :range: 0.0..1.0..4.0
   :feedback: Microphone gain knob

.. mixxx:control:: [MicrophoneN],mute

   Mutes the channel

   :range: binary
   :feedback: Mute button

   .. versionadded:: v2.0.0

.. mixxx:control:: [MicrophoneN],VuMeter

   Outputs the current instantaneous microphone volume

   :range: default
   :feedback: Microphone VU meter changes

   .. versionadded:: v1.10.0

The ``[Recording]`` controls
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. mixxx:control:: [Recording],toggle_recording

   Turns recording on or off.

   :range: binary
   :feedback: Recording icon

.. mixxx:control:: [Recording],status

   Indicates whether Mixxx is currently recording.

   :range:
      ===== ====================
      Value Meaning
      ===== ====================
      0     Recording Stopped
      1     Initialize Recording
      2     Recording Active
      ===== ====================
   :feedback: Recording icon

The ``[AutoDJ]`` controls
~~~~~~~~~~~~~~~~~~~~~~~~~

> 

Reviewer: Maybe add a note/link to the [Library],AutoDJ... controls ?
I know this is just the inital mvoe to the manual, don't wanna demand too much ;)

Author: Done.

You can map audio interface's inputs to mixxx's auxiliary input channels and connect external audio source to it (cellphone, mp3 player).
Then you can use your :term:`MIDI` controller to control its volume and some other parameters (orientation, gain, :term:`volume`), apply effects and use the prelisten function.

.. note:: Although the first auxiliary group is named :mixxx:cogroupref:`[Auxiliary1]`, the group for the first microphone is just called :mixxx:cogroupref:`[Microphone] <[MicrophoneN]>`, not :mixxx:cogroupref:`[Microphone1] <[MicrophoneN]>`.

.. mixxx:control:: [MicrophoneN],enabled
                   [AuxiliaryN],enabled

   1 if a channel input is enabled, 0 if not.

   :range: binary
   :feedback: Microphone is enabled.

   .. versionadded:: v1.10.0

.. mixxx:control:: [MicrophoneN],input_configured
                   [AuxiliaryN],input_configured

   1 if there is input is configured for this channel, 0 if not.

   :range: binary, read-only
   :feedback: Configured channel in the sound preferences.

.. mixxx:control:: [MicrophoneN],master
.. mixxx:control:: [AuxiliaryN],master

   :range: binary

> .. mixxx:control:: [MicrophoneN],orientation
>                    [AuxiliaryN],orientation
> 
>    Set channel orientation,
> 
>    :range:
>       ===== ===================================
>       Value Meaning
>       ===== ===================================
>       0     Left side of crossfader
>       1     Center (not affected by crossfader)
>       2     Right side of crossfader
>       ===== ===================================
>    :feedback: N/A
> 
>    .. versionadded:: v1.10.0
>    .. deprecated:: v1.10.0
> 
>       The control is not processed in the Mixer, which is also why there are no orientation controls for Microphones in the GUI.

Reviewer: orientation works for Aux, let's split off the Mic part

```suggestion
.. mixxx:control:: [AuxiliaryN],orientation

   Set channel orientation,

   :range:
      ===== ===================================
      Value Meaning
      ===== ===================================
      0     Left side of crossfader
      1     Center (not affected by crossfader)
      2     Right side of crossfader
      ===== ===================================
   :feedback: N/A

   .. versionadded:: v1.10.0
      
.. mixxx:control:: [MicrophoneN],orientation

   .. versionadded:: v1.10.0
   .. deprecated:: v1.10.0

      The control is not processed in the Mixer, which is also why there are no orientation controls for Microphones in the GUI.
```

.. mixxx:control:: [MicrophoneN],PeakIndicator
                   [AuxiliaryN],PeakIndicator

   Indicates when the signal is clipping (too loud for the hardware and is being distorted)

   :range: binary
   :feedback: Microphone Clip light

   .. versionadded:: v1.10.0

.. mixxx:control:: [MicrophoneN],PeakIndicatorL
                   [AuxiliaryN],PeakIndicatorL

   Indicates when the signal is clipping (too loud for the hardware and is being distorted) for the left channel

   :range: binary
   :feedback: Clip light (left)

   .. versionadded:: v2.0.0

.. mixxx:control:: [MicrophoneN],PeakIndicatorR
                   [AuxiliaryN],PeakIndicatorR

   Indicates when the signal is clipping (too loud for the hardware and is being distorted) for the right channel

   :range: binary
   :feedback: Clip light (right)

   .. versionadded:: v2.0.0

.. mixxx:control:: [MicrophoneN],pfl
                   [AuxiliaryN],pfl

   Toggles headphone cueing

   :range: binary
   :feedback: Headphone button

> .. mixxx:control:: [MicrophoneN],talkover
>                    [AuxiliaryN],talkover
> 

Reviewer: Let's exclude Aux due to the confusion we just discovered.
```suggestion
.. mixxx:control:: [MicrophoneN],talkover

```

Mixxx Controls
--------------

Nearly every knob, button, or fader you see in Mixxx's interface is
controllable via Mixxx's "control" system. The control system allows

> skins, MIDI controllers, and HID controllers to control Mixxx via a

Reviewer: ```suggestion
skins, MIDI controllers, HID controllers and keyboards to control Mixxx via a
```

single interface.

A control is identified by a "group" (which is used for grouping
associated controls) and a "key" (the name of the individual control).

For example, the volume fader for Deck 1 is identified by the group

> `[Channel1]` and key :mixxx:coref:`volume <[Channel1],volume>`. Similarly, the

Reviewer: why not use :mixxx:coref: for groups, too?
Alos, why

Author: I implemented `:mixxx:cogroupref:` instead.
